src/common/com/intellij/plugins/haxe/ide/index/HaxeSymbolIndex.java

```diff
@@ -0,0 +1,183 @@
+import com.intellij.plugins.haxe.util.HaxeResolveUtil;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.search.GlobalSearchScope;
+import com.intellij.psi.search.PsiElementProcessor;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.ArrayUtil;
+import com.intellij.util.CommonProcessors;
+import com.intellij.util.indexing.*;
+import com.intellij.util.io.EnumeratorStringDescriptor;
+import com.intellij.util.io.KeyDescriptor;
+import gnu.trove.THashMap;
+import gnu.trove.THashSet;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+
+public class HaxeSymbolIndex extends ScalarIndexExtension<String> {
+  public static final ID<String, Void> HAXE_SYMBOL_INDEX = ID.create("HaxeSymbolIndex");
+  private static final int INDEX_VERSION = HaxeIndexUtil.BASE_INDEX_VERSION + 8;
+  private DataIndexer<String, Void, FileContent> myDataIndexer = new MyDataIndexer();
+
+  @NotNull
+  @Override
+  public ID<String, Void> getName() {
+    return HAXE_SYMBOL_INDEX;
+  }
+
+  @Override
+  public int getVersion() {
+    return INDEX_VERSION;
+  }
+
+  @NotNull
+  @Override
+  public DataIndexer<String, Void, FileContent> getIndexer() {
+    return myDataIndexer;
+  }
+
+  @NotNull
+  @Override
+  public KeyDescriptor<String> getKeyDescriptor() {
+    return new EnumeratorStringDescriptor();
+  }
+
+  @NotNull
+  @Override
+  public FileBasedIndex.InputFilter getInputFilter() {
+    return HaxeSdkInputFilter.INSTANCE;
+  }
+
+  @Override
+  public boolean dependsOnFileContent() {
+    return true;
+  }
+
+  public static String[] getAllSymbols(@NotNull final GlobalSearchScope scope) {
+    final CommonProcessors.CollectProcessor<String> processor = new CommonProcessors.CollectProcessor<String>();
+    FileBasedIndex.getInstance().processAllKeys(HAXE_SYMBOL_INDEX, processor, scope, null);
+    return ArrayUtil.toStringArray(processor.getResults());
+  }
+
+  public static List<HaxeComponentName> getItemsByName(@NotNull final String name,
+                                                       @NotNull final Project project,
+                                                       @NotNull final GlobalSearchScope searchScope) {
+    final Collection<VirtualFile> files =
+      FileBasedIndex.getInstance().getContainingFiles(HAXE_SYMBOL_INDEX, name, searchScope);
+    final Set<HaxeComponentName> result = new THashSet<HaxeComponentName>();
+    for (VirtualFile vFile : files) {
+      final PsiFile psiFile = PsiManager.getInstance(project).findFile(vFile);
+      if (psiFile == null || psiFile.getFileType() != HaxeFileType.HAXE_FILE_TYPE) {
+        continue;
+      }
+      processComponents(psiFile, new PsiElementProcessor<HaxeNamedComponent>() {
+        @Override
+        public boolean execute(@NotNull HaxeNamedComponent subComponent) {
+          if (name.equals(subComponent.getName())) {
```

Comment:
> Yes, it works out of the box.

I tested it with the following code:
```
package ;
class Main {
    public function new() {
    }

    var list:List<Int>;

    public function main():Void {
        var list:List;
        list.add("EMB");
    }

    public function add():Void {
        list.clear();
    }

}
```

It works with beginnings:
![image](https://cloud.githubusercontent.com/assets/7231977/13455275/2613b18c-e011-11e5-8977-f4a81e20ccc0.png)

and contiguous letters:
![image](https://cloud.githubusercontent.com/assets/7231977/13455244/edffd21c-e010-11e5-9280-b476454a65c3.png)

but not with split names:
![image](https://cloud.githubusercontent.com/assets/7231977/13455303/57dd25d6-e011-11e5-93dc-81a517ef6cf6.png)

Case insensitive also works:
![image](https://cloud.githubusercontent.com/assets/7231977/13455380/ea7a1692-e011-11e5-850e-91be030d6dff.png)

Even so, it's much better than we had.

Any idea why the local symbols disappear when using discontiguous letters?